File: iOS/Schema/Sources/Schema/Dependencies.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

public protocol DependencySchema {
    var schemaDependencies: SchemaDependencies { get }
}

//extension DependencySchema {
//    public var loggerHelper: SchemaLoggerHelper? {
//        return schemaDependencies.loggerHelper
//    }
//
//    public var decoder: ComponentDecoding {
//        return schemaDependencies.decoder
//    }
//}

typealias SchemaAttributes = DependencyComponentDecoding & DependencyLoggerHelper

public struct SchemaDependencies: SchemaAttributes {
    public var loggerHelper: SchemaLoggerHelper?
    public var decoder: ComponentDecoding

    internal init(
        loggerHelper: SchemaLoggerHelper? = nil,
        decoder: ComponentDecoding
    ) {
        self.loggerHelper = loggerHelper
        self.decoder = decoder
    }

    public init(
        loggerHelper: SchemaLoggerHelper,
        decoder: ComponentDecoding = ComponentDecoder()
    ) {
        self.loggerHelper = loggerHelper
        self.decoder = decoder
    }
    
}

File: tests/appium/app-android/buildSrc/src/main/kotlin/br/com/zup/beagle/Dependencies.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle

import org.gradle.api.Plugin
import org.gradle.api.Project

class Dependencies : Plugin<Project> {
    override fun apply(project: Project) {}

    object Releases {
        const val versionCode = 1
        const val versionName = "1.0"
        const val beagleVersionName = "0.0.1"
    }

    object Versions {
        const val compileSdk = 30
        const val minSdk = 19
        const val targetSdk = 30
        const val buildTools = "30.0.0"
        const val kotlin = "1.4.10"
        const val kotlinCoroutines = "1.3.9"
        const val appcompat = "1.2.0"
        const val viewModel = "2.2.0"
        const val recyclerView = "1.1.0"
        const val junit5 = "5.7.0"
        const val materialDesign = "1.2.1"
        const val gson = "2.8.6"
    }

    object GeneralLibraries {
        const val kotlinCoroutines =
            "org.jetbrains.kotlinx:kotlinx-coroutines-android:${Versions.kotlinCoroutines}"
    }

    object GoogleLibraries {
        const val materialDesign = "com.google.android.material:material:${Versions.materialDesign}"
    }

    object AndroidxLibraries {
        const val appcompat = "androidx.appcompat:appcompat:${Versions.appcompat}"
        const val coreKtx = "androidx.core:core-ktx:${Versions.appcompat}"
    }

    object GsonLibraries{
        const val gson = "com.google.code.gson:gson:${Versions.gson}"
    }


}